File: 102. 二叉树的层序遍历.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
import collections


class Solution(object):
    def levelOrder(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: List[List[int]]
        """
        if root is None:
            return []
        res = []
        q = collections.deque([root])
        while q:
            # 题目要求每一层的数据各自用列表装
            level_size = len(q)
            level_res = []
            for _ in range(level_size):
                curr = q.popleft()
                level_res.append(curr.val)
                # 如果有左右节点，就放到队尾排队
                if curr.left:
                    q.append(curr.left)
                if curr.right:
                    q.append(curr.right)
            res.append(level_res)
        return res

File: 102. 二叉树的层序遍历.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func levelOrder(root *TreeNode) [][]int {
	var res [][]int
	if root == nil {
		return res
	}
	q := make([]*TreeNode, 0)
	q = append(q, root)

	for len(q) != 0 {
		length := len(q)
		currRes := make([]int, 0, length)

		for i := range length {
			currRes = append(currRes, q[i].Val)

			if q[i].Left != nil {
				q = append(q, q[i].Left)
			}
			if q[i].Right != nil {
				q = append(q, q[i].Right)
			}
		}

		q = q[length:]
		res = append(res, currRes)
	}

	return res
}
